packages/client/src/schedule-types.ts

```diff
@@ -0,0 +1,830 @@
+
+export interface WorkflowExecutionWithFirstExecutionRunId {
+  workflowId: string;
+
+  /**
+   * The Run Id of the original execution that was started by the Schedule. If the Workflow retried, did
+   * Continue-As-New, or was Reset, the following runs would have different Run Ids.
+   */
+  firstExecutionRunId: string;
+}
+
+export interface ScheduleAction {
+  /** Time that the Action was scheduled for, including jitter. */
+  scheduledAt: Date;
+
+  /** Time that the Action was actually taken. */
+  takenAt: Date;
+
+  /** If action was {@link StartWorkflowAction}. */
+  workflow?: WorkflowExecutionWithFirstExecutionRunId;
+}
+
+/**
+ * Policy for overlapping Actions.
+ */
+export enum ScheduleOverlapPolicy {
+  /**
+   * Use server default (currently SKIP).
+   *
+   * TODO remove this field if this issue is implemented: https://github.com/temporalio/temporal/issues/3240
+   */
+  UNSPECIFIED = 0,
+
+  /**
+   * Don't start a new Action.
+   */
+  SKIP,
+
+  /**
+   * Start another Action as soon as the current Action completes, but only buffer one Action in this way. If another
+   * Action is supposed to start, but one Action is running and one is already buffered, then only the buffered one will
+   * be started after the running Action finishes.
+   */
+  BUFFER_ONE,
+
+  /**
+   * Allows an unlimited number of Actions to buffer. They are started sequentially.
+   */
+  BUFFER_ALL,
+
+  /**
+   * Cancels the running Action, and then starts the new Action once the cancelled one completes.
+   */
+  CANCEL_OTHER,
+
+  /**
+   * Terminate the running Action and start the new Action immediately.
+   */
+  TERMINATE_OTHER,
+
+  /**
+   * Allow any number of Actions to start immediately.
+   *
+   * This is the only policy under which multiple Actions can run concurrently.
+   */
+  ALLOW_ALL,
+}
+
+checkExtends<
+  keyof typeof temporal.api.enums.v1.ScheduleOverlapPolicy,
+  `SCHEDULE_OVERLAP_POLICY_${keyof typeof ScheduleOverlapPolicy}`
+>();
+checkExtends<
+  `SCHEDULE_OVERLAP_POLICY_${keyof typeof ScheduleOverlapPolicy}`,
+  keyof typeof temporal.api.enums.v1.ScheduleOverlapPolicy
+>();
+
+export interface Backfill {
+  /** Time range to evaluate Schedule in. */
+  start: Date;
+  end: Date;
+
+  /**
+   * Override the Overlap Policy for this request.
+   */
+  overlap?: ScheduleOverlapPolicy;
+}
+
+/**
+ * Example Ranges:
+ *
+ * ```
+ * { start: 2 } ➡️ 2
+ * { start: 2, end: 4 } ➡️ 2, 3, 4
+ * { start: 2, end: 10, step: 3 } ➡️ 2, 5, 8
+ * ```
+ */
+export type Range<Unit> = RangeStartOnly<Unit> | RangeStartEndSkip<Unit>;
+
+interface RangeStartOnly<Unit> {
+  /**
+   * Start of range (inclusive)
+   */
+  start: Unit;
+
+  /**
+   * End of range (inclusive)
+   *
+   * @default `start`
+   */
+  end?: never;
+
+  /**
+   * The step to take between each value.
+   *
+   * @default 1
+   */
+  step?: never;
+}
+
+interface RangeStartEndSkip<Unit> extends Omit<RangeStartOnly<Unit>, 'end' | 'step'> {
```

Comment:
Why do we need this variant?

Reply:
Because we want to support:

```
 * { start: 2 } ➡️ 2
 * { start: 2, end: 4 } ➡️ 2, 3, 4
 * { start: 2, end: 10, step: 3 } ➡️ 2, 5, 8
```

but not:

```
 * { start: 2, step: 1 }
```

So both `end` and `step` are optional, but `step` can't be there unless `end` is also present. I can however shorten this to only the `type` (ie. remove the two useless interfaces).

Reply:
```
export type Range<Unit> =
  | {
      /** Start of range (inclusive) */
      start: Unit;

      /**
       * End of range (inclusive)
       * @default `start`
       */
      end?: Unit;

      /**
       * The step to take between each value.
       * @default 1
       */
      step?: never;
    }
  | {
      /** Start of range (inclusive) */
      start: Unit;

      /** End of range (inclusive) */
      end: Unit;

      /** The step to take between each value. */
      step: number;
    };
```

Reply:
Does it hurt if we have just `start` and `step`?
Isn't it the exact same as just `start`?

Reply:
I think that would be confusing... Intuitively, I would read `{ start: 3, step: 1 }` as meaning an open range starting at 3 and including every other values after... I think it better disallow this form entirely.